### PR TITLE
Added article refactoring

### LINKED ISSUE #261: New article: Refactoring
New article: Refactoring

What is this about: Software needs to be made sleeker, faster, better looking. For this, you have to refactor code - change the inner workings without changing the functionality. Elaborate on the right procedures and rules to follow.

Read more here: https://refactoring.com/

## How to contribute:
- Follow our [Contribution guide](https://github.com/DXHeroes/knowledge-base-content/blob/master/CONTRIBUTING.md)
- How to [Git and Pull Requests]( https://github.com/firstcontributions/first-contributions/blob/master/README.md)
- How to [Markdown](https://guides.github.com/features/mastering-markdown/), [Markdown interactive demo](https://www.markdowntutorial.com/lesson/1/) or [Google Docs Addon](https://gsuite.google.com/marketplace/app/docs_to_markdown/700168918607)

### PR DESCRIPTION
**Pull request by K-Kumar-01**
Closes #261 

## If an article related PR

- [x] I have checked that my suggestion follows the Contribution guide

- [x]  I have added which issue this PR closes

- [ ] I have 0 warnings from Danger.js

- [x] If I have added images, they are in /files/ folder

## If other PR

- [ ] I have added which issue this PR closes

### CODE REVIEW

File: practices/refactoring.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+---
+slug: refactoring
+stages:
+  - development
+  - production
+short_description:In computer programming and software design, code refactoring is the process of restructuring existing computer code—changing the factoring—without changing its external behavior.
```
Comment (prokopsimek) [inline practices/refactoring.md:6]:
```suggestion
short_description: In computer programming and software design, code refactoring is the process of restructuring existing computer code—changing the factoring—without changing its external behavior.
```

File: practices/refactoring.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+---
+slug: refactoring
+stages:
+  - development
+  - production
```
Comment (prokopsimek) [inline practices/refactoring.md:5]:
```suggestion
  - development
```

`production` is not a valid stage - see Contribution guide for more info

**Review by prokopsimek — CHANGES_REQUESTED** [review]
@K-Kumar-01 please see my suggestions. There could be more paragraphs but the article is fine.

cc @ridlees

**Comment by K-Kumar-01** [comment]
@prokopsimek  I have made the suggested changes.

**Review by prokopsimek — APPROVED** [review]

File: practices/refactoring.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+---
+slug: refactoring
+stages:
+  - development
+short_description: In computer programming and software design, code refactoring is the process of restructuring existing computer code—changing the factoring—without changing its external behavior.
```
Comment (ridlees) [inline practices/refactoring.md:5]: Please, rewrite this definition with your own words. Using the direct quote (even though is a good one) hurts this article.

_In computer programming and software design, code refactoring is the process of restructuring existing computer code—changing the factoring—without changing its external behavior_ from https://en.wikipedia.org/wiki/Code_refactoring

File: practices/refactoring.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+- Bug introduction is quite common.
+- Does not mean rewriting the code.
+- Does not mean fixing the bugs
+
+## Resources for Refactoring
+
+- [Refactoring Guru](https://refactoring.guru/refactoring)
+- [Wikipedia](https://en.wikipedia.org/wiki/Code_refactoring)
+- [Science Direct](https://www.sciencedirect.com/topics/computer-science/refactoring)
+- [Agile Alliance](https://www.agilealliance.org/glossary/refactoring/#q=~(infinite~false~filters~(postType~(~'page~'post~'aa_book~'aa_event_session~'aa_experience_report~'aa_glossary~'aa_research_paper~'aa_video)~tags~(~'refactoring))~searchTerm~'~sort~false~sortDirection~'asc~page~1))
```
Comment (ridlees) [inline practices/refactoring.md:82]:
```suggestion
- [Agile Alliance](https://www.agilealliance.org/glossary/refactoring)
```

File: practices/refactoring.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+## Common pitfalls of Refactoring
+
+- Challenging as small changes are not likely to be made because of laziness i.e. "Too small to worth doing".
+- Expensive and risky in view of management.
+- Bug introduction is quite common.
+- Does not mean rewriting the code.
+- Does not mean fixing the bugs
+
+## Resources for Refactoring
+
```
Comment (ridlees) [inline practices/refactoring.md:78]: I think you could also include this paper on the correlation between refactoring as practice and quality of code (10.1051/matecconf/20165702012) (if you don't have access to the academic sources, it is possible to use  https://scihub.wikicn.top to get the content. I am not encouraging you to use it).

File: practices/refactoring.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+- Code Debugging is improved.
+- Improves Code smell.
+- Improved readability.
+- Reduced complexity.
+- Helps code understanding.
+- Allows reusable design elements.
+
+## How to Implement Refactoring
+
+1. Find the code which is used in many places.
```
Comment (ridlees) [inline practices/refactoring.md:57]:
```suggestion
1. Find the code which you use in many places.
```

File: practices/refactoring.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+2. Create its own function.
+3. Call that function in the place.
```
Comment (ridlees) [inline practices/refactoring.md:59]:
```suggestion
3. Call that function in the place instead.
```

File: practices/refactoring.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+4. Repeat till steps 1-3 cannot be done.
```
Comment (ridlees) [inline practices/refactoring.md:60]:
```suggestion
4. Repeat till you can not find frequently used code
```

File: practices/refactoring.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+**Some things to be taken care of **
```
Comment (ridlees) [inline practices/refactoring.md:62]:
```suggestion
**Keep in mind during refactoring**
```

File: practices/refactoring.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+> A good program is as good as its code.
+
+If the code of a program is easily readable and understandable then the program made by it is better as compared to a code which is messy.
+
+The essence of refactoring is applying small coding changes that preserves the functionality. The small changes when combined together bring about a significant change.
+
+## Why You Might Want to Refactor?
+
+- Refactoring the code makes it easier to understand for other people.
```
Comment (ridlees) [inline practices/refactoring.md:38]:
```suggestion
- Refactoring your code makes it easier to understand for other people.
```

File: practices/refactoring.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+- It avoids the duplication and hence one follows DRY principle
+- It reduces the code which is less maintenance.
```
Comment (ridlees) [inline practices/refactoring.md:40]:
```suggestion
- It reduces code maintenance
```

File: practices/refactoring.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+- It passes all the tests.
```
Comment (ridlees) [inline practices/refactoring.md:41]:
```suggestion
- It makes testing easier
```
I think that any piece of code should pass all the tests (not only refactored).

File: practices/refactoring.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+- Debugging and maintenance is easier.
```
Comment (ridlees) [inline practices/refactoring.md:42]:
```suggestion
- Debugging is easier.
```

File: practices/refactoring.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+## Problems the Refactoring Helps to Solve
+
+- Improves the objective attributes of code(length, duplication, coupling ,etc.)
```
Comment (ridlees) [inline practices/refactoring.md:46]: unfortunately, this sentence is too similar to _refactoring improves objective attributes of code (length, duplication, coupling, and cohesion_

File: practices/refactoring.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+- Maintenance is much easier.
+- Code Debugging is improved.
+- Improves Code smell.
+- Improved readability.
+- Reduced complexity.
+- Helps code understanding.
+- Allows reusable design elements.
+
+## How to Implement Refactoring
+
```
Comment (ridlees) [inline practices/refactoring.md:56]: I would suggest mentioning something about a correct version control and refactoring. Something like when committing, use the label “Refactoring”. What do you think about this suggestion? Do you agree?

File: practices/refactoring.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+- Improves the objective attributes of code(length, duplication, coupling ,etc.)
+- Maintenance is much easier.
+- Code Debugging is improved.
+- Improves Code smell.
+- Improved readability.
+- Reduced complexity.
+- Helps code understanding.
+- Allows reusable design elements.
+
```
Comment (ridlees) [inline practices/refactoring.md:54]:
```suggestion
- Helps to minimize [Poor Code Quality](/problems/poor-code-quality)
```
Some of the other problems could be included in this. What do you think about it?

File: practices/refactoring.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+- If code is not cleaner refactoring is not done.
+- Don't include new features.
+- Try making small changes step by step. 
+- All tests should pass
+
+## Common pitfalls of Refactoring
+
+- Challenging as small changes are not likely to be made because of laziness i.e. "Too small to worth doing".
+- Expensive and risky in view of management.
```
Comment (ridlees) [inline practices/refactoring.md:72]:
```suggestion
- Expensive and risky in the view of management.
```
Or you could say Management can see refactoring as expensive and not important. Just suggestion

File: practices/refactoring.md [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+- Bug introduction is quite common.
+- Does not mean rewriting the code.
```
Comment (ridlees) [inline practices/refactoring.md:74]:
```suggestion
- Developer start adding new features, rewriting the code, or fixing bugs during refactoring.
```
I think these pitfalls can be grouped together.

File: practices/refactoring.md [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+**Some things to be taken care of **
+
+- If code is not cleaner refactoring is not done.
+- Don't include new features.
+- Try making small changes step by step. 
+- All tests should pass
+
+## Common pitfalls of Refactoring
+
+- Challenging as small changes are not likely to be made because of laziness i.e. "Too small to worth doing".
```
Comment (ridlees) [inline practices/refactoring.md:71]:
```suggestion
- Lack of developer motivation to do these small changes
```

**Review by ridlees — CHANGES_REQUESTED** [review]
Hello @K-Kumar-01, sorry for my delay, I had some sick days. I like how on point and fresh your content is. Good job! Please, look at my suggestion and let me know if you find them valuable.

**Review by ridlees — APPROVED** [review]

**Comment by ridlees** [comment]
Hello @K-Kumar-01, since it is a great article that I like a lot, I have taken the liberty to address all the comments. I am glad that we can publish it. Hurray 🎉 🎉 🎉

**Pull request merged**